frontend/src/components/MealForm.js

```diff
@@ -0,0 +1,269 @@
+import React, { useState, useEffect } from "react";
+import { useDispatch } from "react-redux";
+import { me } from "../store/authSlice";
+import Autocomplete from "./Autocomplete";
+import apiClient from "../config";
+import { Grid, Row, Spacer, Text, Col } from "@nextui-org/react";
+import { Button, Table, Inputs, HeaderText } from "./nextUI/index";
+import SuccessMessage from "./SuccessMessage";
+
+//form for entering a meal
+function MealForm() {
+  const [date, setDate] = useState("");
+  const [time, setTime] = useState("");
+  const [entryName, setEntryName] = useState("");
+  const [mealArray, setMealArray] = useState([]);
+  const [success, setSuccess] = useState(false);
+  const [error, setError] = useState("");
+  const [allGroups, setAllGroups] = useState([]);
+  const [recentFoods, setRecentFoods] = useState([]);
+  const [loading, setLoading] = useState(false);
+
+  const dispatch = useDispatch();
+
+  // Dispatch to get user information
+  useEffect(() => {
+    dispatch(me());
+  }, [dispatch]);
+
+  //fetch food groups when the component mounts
+  useEffect(() => {
+    const today = new Date().toISOString();
+    setDate(today.substring(0, 10));
+
+    fetchGroups();
+    fetchRecentFoods();
+  }, []);
+
+  //fetch groups from api
+  const fetchGroups = async () => {
+    try {
+      const { data } = await apiClient.get("groups/");
+
+      if (data && data["data"]) {
+        setAllGroups(data["data"]);
+      }
+    } catch (err) {
+      console.error(err);
+    }
+  };
+
+  //fetch the user's recent foods
+  const fetchRecentFoods = async () => {
+    try {
+      const { data } = await apiClient.get("meals/recent");
+
+      //[{name, groups}, {name, groups}, ...]
+      if (data) setRecentFoods(data);
+    } catch (err) {
+      console.error(err);
+    }
+  };
+
+  //add a food to the meal array
+  const addFood = (foodObj) => {
+    setError("");
+    setMealArray([...mealArray, foodObj]);
+  };
+
+  //handle form changes
+  const handleDateChange = (event) => {
+    setDate(event.target.value);
+  };
+
+  const handleTimeChange = (event) => {
+    setTime(event.target.value);
+  };
+
+  const handleNameChange = (event) => {
+    setEntryName(event.target.value);
+  };
+
+  //filter for unique values (helper function used before submitting the meal)
+  const makeUnique = (array) => {
+    return array.filter((item, idx) => array.indexOf(item) === idx);
+  };
+
+  //submit the meal
+  const handleSubmit = async (e) => {
+    try {
+      e.preventDefault();
+
+      //set an error if the user hasn't added a food
+      if (!mealArray.length) {
+        setError("Please add at least one food");
+        return;
+      }
+
+      setLoading(true);
+
+      //filter and format the foods and groups for mongoDB
+      const foods = makeUnique(mealArray.map(({ name }) => name));
+      const groups = mealArray.reduce((array, { groups }) => {
```

Comment:
Seems clunky to use `forEach` and `push` inside `reduce` - you might be able to do the same thing with a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)